lib/key_manager.rb

```diff
@@ -0,0 +1,71 @@
+require 'securerandom'
+require 'thread'
+
+class KeyManager
+  attr_reader :keys, :blocked_keys
+
+  def initialize
+    @keys = {}
+    @blocked_keys = {}
+    @mutex = Mutex.new
+  end
+
+  def generate_key
+    key = SecureRandom.hex(16)
+    # 5 min expiry
+    expiry = Time.now + 300 
+    @mutex.synchronize do
+      @keys[key] = expiry
+    end
+    key
+  end
+
+  def get_available_key
+    @mutex.synchronize do
+      now = Time.now
+      available_keys = @keys.reject { |_, expiry| expiry < now }
+      unblocked_keys = available_keys.reject { |key, _| @blocked_keys.key?(key) }
+
+      return nil if unblocked_keys.empty?
+
+      key, _ = unblocked_keys.to_a.sample
+      # block this key for 60 seconds
+      @blocked_keys[key] = now + 60 
```

Comment:
Here, for example
1 key was generated with Expiry of 5 at time 0
Then at time 1 it was blocked and you would add the value as 2 in blocked keys value
Then after 2 whenever your cleanup runs, it will delete the key but once key is unblocked the expiry time is not updated.

Reply:
Have added the logic for the same